### DISCUSSION TITLE
Use one schema for each user

### DISCUSSION THREAD

**Discussion by guiqis** [prisma/prisma#25287] [Q&A]
### Question

Hi guys! 

I'm considering using one schema for each of my users. Each user has a unique ID and the schema will be something like: user_{id}. How can I set prisma in a way that the query will go to the right schema? I found a few solutions using one connection for each user, but I don't think that's a good alternative. If I have 200 users using their accounts that will be 200 connections.

Does anyone have an tip or tutorial on how to solve it?

(All schemas will have the same tables/structure).

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @guiqis 👋 

Would [overriding the datasourceUrl](https://www.prisma.io/docs/orm/reference/prisma-client-reference#programmatically-override-a-datasource-url) property be a viable approach for routing queries to different schemas or databases based on user-specific logic?

Since each user in your case has their own schema (e.g., user_{id}), you can dynamically set the connection string at runtime depending on which user is making the request. You can programmatically change the `datasourceUrl` (the database connection string) for each query to target the correct schema, without modifying the `schema.prisma` file or creating multiple connections.

**Reply by guiqis** [reply]
Hey! Sorry for my late response. I'll give it a try. But one question: how can I set the user ID in the connection, since prisma initialization cannot be inside a function? Will I need to "create" a new connection for all queries?

**Reply by ludralph** [reply]
Hi,

You could create a custom wrapper around Prisma that manages schema selection. This might involve creating a new Prisma client for each request or using a pool of pre-initialized clients. Here's a conceptual example:
```ts
import { PrismaClient } from '@prisma/client';

class MultiTenantPrisma {
  private clients: Map<string, PrismaClient> = new Map();

  getClientForUser(userId: string) {
    if (!this.clients.has(userId)) {
      const databaseUrl = `postgresql://user:password@localhost:5432/database?schema=user_${userId}`;
      this.clients.set(userId, new PrismaClient({ datasources: { db: { url: databaseUrl } } }));
    }
    return this.clients.get(userId);
  }
}
```
// Usage
```ts
const multiTenantPrisma = new MultiTenantPrisma();
const userPrisma = multiTenantPrisma.getClientForUser('123');
const result = await userPrisma.someModel.findMany();
```
This approach still creates multiple Prisma clients, but it manages them centrally.

For more control, you could use Prisma's `$executeRaw` or `$queryRaw` methods to explicitly set the schema for each query:
```js
const result = await prisma.$executeRaw`SET search_path TO user_${userId}; SELECT * FROM some_table;`;
```
However, this approach loses many of Prisma's benefits like type safety and query building.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
